### PR TITLE
Brings VFD SWMR code closer to THG standards and allows C library compilation on Windows

### PR DESCRIPTION
**Pull request by derobins**
These changes bring the VFD SWMR branch into closer agreement with THG coding standards and allow the main library to be built on Windows.

### CODE REVIEW

File: MANIFEST [hunk 1]
```diff
@@ -521,8 +528,6 @@
 ./src/hdf5.lnt        _DO_NOT_DISTRIBUTE_
 ./src/hdf5-win.lnt      _DO_NOT_DISTRIBUTE_
 ./src/hdf5-lin.lnt      _DO_NOT_DISTRIBUTE_
-./src/hlog.c
-./src/hlog.h
```
Comment (derobins) [inline MANIFEST:525]: The logging layer uses POSIX things that are not present on Windows and most of it looked like printf debugging, so I ripped it out.

File: config/cmake/H5pubconf.h.in [hunk 2]
```diff
@@ -222,6 +222,9 @@
 /* Define to 1 if you have the `dmalloc' library (-ldmalloc). */
 #cmakedefine H5_HAVE_LIBDMALLOC @H5_HAVE_LIBDMALLOC@
 
+/* Define to 1 if you have the <libgen.h> header file. */
+#cmakedefine H5_HAVE_LIBGEN_H @H5_HAVE_LIBGEN_H@
+
```
Comment (derobins) [inline config/cmake/H5pubconf.h.in:227]: libgen.h is used for basename and dirname, which are only used in the tests. This stuff may go away due to differing semantics between them and their Win32 equivalents.

File: src/H5FDvfd_swmr.c [hunk 3]
```diff
@@ -269,14 +263,14 @@ static herr_t
...
-    bool do_try;                             /* more tries remain */
+    hbool_t do_try;                         /* more tries remain */
     herr_t      ret_value = SUCCEED;
     FUNC_ENTER_STATIC
 
     file->api_elapsed_nbuckets = file->config.max_lag + 1;
 
     file->api_elapsed_ticks =
-        calloc(file->api_elapsed_nbuckets, sizeof(*file->api_elapsed_ticks));
+        H5MM_calloc(file->api_elapsed_nbuckets * sizeof(*file->api_elapsed_ticks));
```
Comment (derobins) [inline src/H5FDvfd_swmr.c:274]: Not all memory allocation has been moved to H5MM. Doing so exposes a memory issue that I haven't corrected yet. This causes failures in debug builds when the H5MM sanity checks complain.

File: src/H5Fvfd_swmr.c [hunk 4]
```diff
@@ -1614,10 +1502,14 @@ H5F__vfd_swmr_update_end_of_tick_and_tick_num(H5F_shared_t *shared,
...
-        HGOTO_ERROR(H5E_FILE, H5E_CANTGET, FAIL, \
-                    "can't get time via clock_gettime")
+#ifdef H5_HAVE_WIN32_API
+        if (timespec_get(&curr, TIME_UTC) != TIME_UTC)
+            HGOTO_ERROR(H5E_FILE, H5E_CANTGET, FAIL, "can't get time via timespec_get");
+#else
+        if (HDclock_gettime(CLOCK_MONOTONIC, &curr) < 0) {
+            HGOTO_ERROR(H5E_FILE, H5E_CANTGET, FAIL, "can't get time via clock_gettime");
+        }
+#endif
```
Comment (derobins) [inline src/H5Fvfd_swmr.c:1512]: This isn't the best protection as clock_gettime(3) isn't (easily) available on some POSIX-y platforms. I'll eventually move this code into a function in H5system.c or H5timer.c when I fix the Windows issues in the tests.

File: src/H5Fvfd_swmr.c [hunk 5]
```diff
@@ -2033,43 +1920,29 @@ H5F__vfd_swmr_writer__wait_a_tick(H5F_t *f)
...
     HDassert(shared->vfd_swmr_writer);
 
     tick_in_nsec = shared->vfd_swmr_config.tick_len * nanosecs_per_tenth_sec;
-    req.tv_nsec = (long)(tick_in_nsec % nanosecs_per_second);
-    req.tv_sec = (time_t)(tick_in_nsec / nanosecs_per_second);
-
-    result = HDnanosleep(&req, &rem);
 
-    while ( result == -1 ) {
+    H5_nanosleep(tick_in_nsec);
```
Comment (derobins) [inline src/H5Fvfd_swmr.c:1935]: H5_nanosleep() will eventually move to the tests, too.

File: src/H5MF.c [hunk 6]
```diff
@@ -134,58 +133,38 @@ hbool_t H5_PKG_INIT_VAR = FALSE;
...
-HLOG_OUTLET_SHORT_DEFN(h5mf_free, h5mf);
-HLOG_OUTLET_SHORT_DEFN(h5mf_alloc, h5mf);
-HLOG_OUTLET_MEDIUM_DEFN(noisy_h5mf_alloc, h5mf_alloc, HLOG_OUTLET_S_OFF);
-HLOG_OUTLET_SHORT_DEFN(h5mf_extend, h5mf);
-HLOG_OUTLET_SHORT_DEFN(h5mf_shrink, h5mf);
-
 static herr_t
-defer_free(H5F_shared_t *shared, H5FD_mem_t alloc_type, haddr_t addr,
+H5MF__defer_free(H5F_shared_t *shared, H5FD_mem_t alloc_type, haddr_t addr,
     hsize_t size)
```
Comment (derobins) [inline src/H5MF.c:138]: Fixing all the new functions to be like other HDF5 internal API calls is a work in progress.

File: src/H5win32defs.h [hunk 7]
```diff
@@ -169,11 +168,11 @@ extern "C" {
     H5_DLL char* Wgetlogin(void);
     H5_DLL int c99_snprintf(char* str, size_t size, const char* format, ...);
     H5_DLL int c99_vsnprintf(char* str, size_t size, const char* format, va_list ap);
-    H5_DLL int Wnanosleep(const struct timespec *req, struct timespec *rem);
     H5_DLL herr_t H5_expand_windows_env_vars(char **env_var);
     H5_DLL const wchar_t *H5_get_utf16_str(const char *s);
     H5_DLL int Wopen_utf8(const char *path, int oflag, ...);
     H5_DLL int Wremove_utf8(const char *path);
+    H5_DLL int H5_get_win32_times(H5_timevals_t* tvs);
```
Comment (derobins) [inline src/H5win32defs.h:175]: This is probably missing in develop, too. I'll have a PR up for a few things I noticed while working on the Windows code.

File: test/vfd_swmr.c [hunk 8]
```diff
@@ -1237,6 +1231,10 @@ test_reader_md_concur(void)
         if(HDclose(child_pfd[0]) < 0)
             HDexit(EXIT_FAILURE);
 
+        /* Free unused configuration */
+        if(config_writer)
+            HDfree(config_writer);
+
```
Comment (derobins) [inline test/vfd_swmr.c:1237]: This fixes a memory leak I spotted when running the test through valgrind.

File: test/page_buffer.c [hunk 9]
```diff
@@ -20,7 +20,6 @@
 *************************************************************/
 
 #include <err.h>
```
Comment (derobins) [inline test/page_buffer.c:22]: I left err.h in place when I removed the redundant C/POSIX library includes as ripping out the err.h-based error handling is next on my list.

File: MANIFEST (hunk 1 above)
Comment (gnuoyd) [reply MANIFEST:525]: I ask that you please leave the hlog definitions and statements in place.

If you do not have time to make hlog work on Windows, I understand. But it's rather easy to provide definitions for Windows that turn every hlog definition (`HLOG_OUTLET_SHORT_DEFN(x, y)`) and statement (`hlog_fast(x, ...)`) into a no-op.

The hlog_fast statements and the log outlets that are in VFD SWMR, today, were crafted to diagnose bugs that I was hunting. They were indispensable, later, when I needed to diagnose other bugs. That resembles my experience of using hlog in various projects over the last 18 years: to add the logging that revealed the bug I was working on took some time, but once the logging was in place, it paid for itself over and over again. To write and delete log statements, re-write and re-delete them, by contrast, is tedious and, over the lifetime of a typical development project, very costly.

BTW, I am using hlog in the Mochi project, for which I added an experimental new feature, per-thread ring buffers.

File: src/H5Fvfd_swmr.c [hunk 10]
```diff
@@ -2033,43 +1920,29 @@ H5F__vfd_swmr_writer__wait_a_tick(H5F_t *f)
...
     struct timespec rem;
     uint64_t tick_in_nsec;
-    H5F_shared_t *shared = f->shared;
+    H5F_shared_t *shared;
     herr_t ret_value = SUCCEED;              /* Return value */
 
     FUNC_ENTER_STATIC
 
+    HDassert(f);
+    shared = f->shared;
```
Comment (gnuoyd) [inline src/H5Fvfd_swmr.c:1929]: What are you trying to do by asserting that `f` is not NULL? I expect for the very next statement to trap if `f` is NULL on every platform we support. Further, I expect it to trap for many non-NULL dangling pointers.

File: MANIFEST (hunk 1 above)
Comment (gnuoyd) [reply MANIFEST:525]: Here are my belated comments. Apparently I clicked "Start a review" and that queues comments to submit in a batch.

Comment (derobins) [reply MANIFEST:525]: We can revisit this later. I'd be excited about adding a logging layer to HDF5. I'm not so excited about a logging system that is non-portable and implemented ina scattershot manner.

**Pull request merged**